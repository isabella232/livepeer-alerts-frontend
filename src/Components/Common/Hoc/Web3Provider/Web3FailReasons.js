export const NO_WEB3 = 'The user does not have web3 injected'
export const NO_PERMISSIONS = 'The user did not gave us permissions'
export const WRONG_NETWORK = 'The user is not logged on the correct network'
