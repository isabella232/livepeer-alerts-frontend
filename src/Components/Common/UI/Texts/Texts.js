export const NO_WEB3 = 'Please install metamask and come back!'
export const NO_PERMISSIONS = 'You need to accept the permissions to use our app!'
export const WRONG_NETWORK =
  'You are on the wrong network, you need to use ... in order to use our app!'
export const DEFAULT_ERROR = 'Something went wrong!'
