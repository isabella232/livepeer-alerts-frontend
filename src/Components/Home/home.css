.Home{

}

.Toast {
    color: red;
    background-color: inherit;
}

.Toast-progress-bar {
    background: inherit;
}