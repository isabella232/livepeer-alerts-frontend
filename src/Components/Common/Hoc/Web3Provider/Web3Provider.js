import React, { Component } from 'react'
import Web3 from 'web3'
import * as failReasons from './Web3FailReasons'
import * as texts from '../../UI/Texts/Texts'
import Spinner from '../../UI/Spinner/Spinner'
import logger from '../../../../utils'

const defaultState = {
  userData: 1,
  authenticated: 2,
  web3: 3,
  displayMsg: '',
  error: false
}

const Web3Context = React.createContext(defaultState)

export const Web3ContextConsumer = Web3Context.Consumer

class Web3Provider extends Component {
  state = {
    web3: null,
    userData: {
      authenticated: false,
      reason: null,
      address: null,
      currentNetwork: null
    },
    render: false,
    requestingAuth: true,
    displayMsg: texts.WAITING_AUTHORIZATION,
    error: false
  }

  loadWeb3 = async () => {
    logger.log('[Web3FunctionalProvider.js] loadWeb3')
    /** We check if metamask is installed, either the new version or the legacy one **/
    if (window.ethereum) {
      await this.loadWeb3LastVersion()
    } else if (window.web3) {
      await this.loadWeb3Legacy()
    } else {
      /** The user does not have web3 **/
      logger.log('[Web3FunctionalProvider.js] user does not have web3')
      this.setState({
        render: true,
        userData: {
          authenticated: false,
          reason: failReasons.NO_WEB3
        },
        requestingAuth: false,
        displayMsg: texts.NO_WEB3,
        error: true
      })
    }
  }

  loadWeb3Legacy = async () => {
    logger.log('[Web3FunctionalProvider.js] getting web3 legacy instance')
    let web3Instance
    web3Instance = new Web3(window.web3.currentProvider)
    await this.loadUserDataFromWeb3(web3Instance)
  }

  loadWeb3LastVersion = async () => {
    logger.log('[Web3FunctionalProvider.js] getting web3 new instance')
    let web3Instance
    web3Instance = new Web3(window.ethereum)
    try {
      /** Request access to the user **/
      logger.log('[Web3FunctionalProvider.js] requesting user permissions')
      let enabledAddress = await window.ethereum.enable()
      /** Could be deleted, this is done for make easier the test **/
      if (!enabledAddress) {
        throw new Error(failReasons.NO_PERMISSIONS)
      }
      logger.log('[Web3FunctionalProvider.js] user with web3 ethereum authenticated')
      /** The user accepted the app, now it's authenticated **/
      await this.loadUserDataFromWeb3(web3Instance)
    } catch (error) {
      /** The user denied the app, it's not authenticated **/
      logger.log('[Web3FunctionalProvider.js] user denied access')
      this.setState({
        web3: web3Instance,
        userData: {
          authenticated: false,
          reason: failReasons.NO_PERMISSIONS
        },
        render: true,
        requestingAuth: false,
        displayMsg: texts.NO_PERMISSIONS,
        error: true
      })
      logger.log('[Web3FunctionalProvider.js] user with ethereum denied the access')
    }
  }

  loadUserDataFromWeb3 = async web3Instance => {
    console.log('Loading user data from web3')
    let userAddress
    let userNetwork
    Promise.all([web3Instance.eth.getAccounts(), web3Instance.eth.net.getId()]).then(results => {
      userAddress = results[0]
      userNetwork = results[1]
      web3Instance.eth.getBalance(userAddress[0]).then(balance => {
        /** Convert balance to eth **/
        balance = web3Instance.utils.fromWei(balance, 'ether')
        this.setState({
          web3: web3Instance,
          userData: {
            authenticated: true,
            address: this.toChecksumAddress(web3Instance, userAddress[0]),
            currentNetwork: userNetwork,
            ethBalance: balance
          },
          render: true,
          requestingAuth: false
        })
        /** We subscribe to the event that detects if the user has changed the account **/
        window.ethereum.on('accountsChanged', accounts => {
          logger.log('ETHEREUM CHANGED')
          logger.log('ACCOUNTS ', accounts)
          this.setState({
            userData: {
              ...this.state.userData,
              address: this.toChecksumAddress(web3Instance, accounts[0])
            }
          })
        })
        /** We subscribe to the event that detects if the user has changed the network **/
        window.ethereum.on('networkChanged', network => {
          logger.log('NETWORK CHANGED')
          logger.log('NETWORK ', network)
          this.setState({
            userData: {
              ...this.state.userData,
              currentNetwork: network
            }
          })
        })
      })
    })
  }

  async componentDidMount() {
    logger.log('[Web3FunctionalProvider.js] componentDidMount')
    this.setState({ render: true })
    await this.loadWeb3()
  }

  /** Converts the address from uppercase to lowercase (checksum format) in order to avoid metamask bug of using both address **/
  toChecksumAddress = (web3, address) => {
    let checksumAddress = '0x'
    address = address.toLowerCase().replace('0x', '')

    // creates the case map using the binary form of the hash of the address
    let caseMap = parseInt(web3.utils.sha3('0x' + address), 16)
      .toString(2)
      .substring(0, 40)

    for (let i = 0; i < address.length; i++) {
      if (caseMap[i] === '1') {
        checksumAddress += address[i].toUpperCase()
      } else {
        checksumAddress += address[i]
      }
    }
    return checksumAddress
  }

  render() {
    let content = (
      <>
        <h3>{this.state.displayMsg}</h3>
        <Spinner />
      </>
    )
    if (!this.state.requestingAuth) {
      if (this.state.render && this.state.web3 && this.state.userData.authenticated) {
        content = (
          <Web3Context.Provider
            value={{
              web3: this.state.web3,
              authenticated: this.state.userData.authenticated,
              userData: this.state.userData
            }}
          >
            {this.props.children}
          </Web3Context.Provider>
        )
      } else {
        content = (
          <Web3Context.Provider
            value={{
              web3: this.state.web3,
              authenticated: this.state.userData.authenticated,
              userData: this.state.userData,
              displayMsg: this.state.displayMsg,
              error: this.state.error
            }}
          >
            {this.props.children}
          </Web3Context.Provider>
        )
      }
    }
    return content
  }
}

export default Web3Provider
