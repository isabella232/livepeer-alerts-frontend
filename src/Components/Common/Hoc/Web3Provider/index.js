import * as failReasons from './Web3FailReasons'
import * as texts from '../../Texts'
import React, { Component } from 'react'
import Web3 from 'web3'
import logdown from 'logdown'
import FullLoading from '../../FullLoading'

const logger = logdown('Livepeer:Web3Provider')
logger.state.isEnabled = process.env.NODE_ENV !== 'production'

const defaultState = {
  web3: null,
  userData: {
    authenticated: false,
    reason: null,
    address: null,
    currentNetwork: null,
  },
  render: false,
  requestingAuth: false,
  displayMsg: texts.WAITING_AUTHORIZATION,
  error: false,
}

const Web3Context = React.createContext(defaultState)

export const Web3ContextConsumer = Web3Context.Consumer

class Web3Provider extends Component {
  state = defaultState

  loadWeb3 = async () => {
    logger.log('LoadWeb3')
    // We check if metamask is installed, either the new version or the legacy one
    if (window.ethereum) {
      await this.loadWeb3LastVersion()
    } else if (window.web3) {
      await this.loadWeb3Legacy()
    } else {
      // The user does not have web3
      logger.log('User does not have web3')
      throw new Error(texts.NO_WEB3)
    }
  }

  loadWeb3Legacy = async () => {
    logger.log('Getting web3 legacy instance')
    let web3Instance
    web3Instance = new Web3(window.web3.currentProvider)
    await this.loadUserDataFromWeb3(web3Instance)
  }

  loadWeb3LastVersion = async () => {
    logger.log('Getting web3 new instance')
    let web3Instance
    web3Instance = new Web3(window.ethereum)
    try {
      // Request access to the user
      logger.log('Requesting user permissions')
      let enabledAddress = await window.ethereum.enable()
      // Could be deleted, this is done for make easier the test
      if (!enabledAddress) {
        throw new Error(failReasons.NO_PERMISSIONS)
      }
      logger.log('User with web3 ethereum authenticated')
      // The user accepted the app, now it's authenticated
      await this.loadUserDataFromWeb3(web3Instance)
    } catch (error) {
      // The user denied the app, it's not authenticated
      logger.log('User with ethereum denied the access')
      this.setErrorState(failReasons.NO_PERMISSIONS)
      throw new Error(texts.NO_PERMISSIONS)
    }
  }

  setErrorState(errorReason, web3Instance = null) {
    switch (errorReason) {
      case failReasons.NO_ADDRESS: {
        logger.log('User address not found')
        this.setState({
          web3: web3Instance,
          userData: {
            authenticated: false,
            reason: failReasons.NO_ADDRESS,
          },
          render: true,
          requestingAuth: false,
          displayMsg: texts.NO_ADDRESS,
          error: true,
        })
        break
      }
      case failReasons.NO_PERMISSIONS: {
        this.setState({
          web3: web3Instance,
          userData: {
            authenticated: false,
            reason: failReasons.NO_PERMISSIONS,
          },
          render: true,
          requestingAuth: false,
          displayMsg: texts.NO_PERMISSIONS,
          error: true,
        })
        break
      }
      case failReasons.NO_WEB3: {
        this.setState({
          render: true,
          userData: {
            authenticated: false,
            reason: failReasons.NO_WEB3,
          },
          requestingAuth: false,
          displayMsg: texts.NO_WEB3,
          error: true,
        })
        break
      }
      default: {
        this.setState({
          render: true,
          userData: {
            authenticated: false,
            reason: failReasons.NO_REASON,
          },
          requestingAuth: false,
          displayMsg: texts.DEFAULT_ERROR,
          error: true,
        })
        break
      }
    }
  }

  loadUserDataFromWeb3 = async web3Instance => {
    logger.log('Loading user data from web3')
    let userAddress
    let userNetwork

    const [ethAccounts, networkId] = await Promise.all([web3Instance.eth.getAccounts(), web3Instance.eth.net.getId()])
    userAddress = ethAccounts
    userNetwork = networkId
    logger.log('User address', userAddress)
    logger.log('User network ', userNetwork)

    // Once we got the user data, we check if he has an address, otherwise we should throw an error
    if (userAddress.length === 0) {
      this.setErrorState(failReasons.NO_ADDRESS, web3Instance)
      return
    }

    let balance = await web3Instance.eth.getBalance(userAddress[0])
    // Convert balance to eth
    balance = web3Instance.utils.fromWei(balance, 'ether')
    this.setState({
      web3: web3Instance,
      userData: {
        authenticated: true,
        address: this.toChecksumAddress(web3Instance, userAddress[0]),
        currentNetwork: userNetwork,
        ethBalance: balance,
      },
      render: true,
      requestingAuth: false,
    })

    // We subscribe to the event that detects if the user has changed the account
    window.ethereum.on('accountsChanged', accounts => {
      window.location.reload()
    })
    // We subscribe to the event that detects if the user has changed the network
    window.ethereum.on('networkChanged', network => {
      return this.networkChangedCallBack(web3Instance, network, balance)
    })
  }

  accountChangedCallback = (web3Instance, accounts) => {
    logger.log('Account Changed ', accounts)
    this.getUserBalance(web3Instance, accounts[0]).then(balance => {
      balance = web3Instance.utils.fromWei(balance, 'ether')
      this.setState({
        userData: {
          ...this.state.userData,
          address: this.toChecksumAddress(web3Instance, accounts[0]),
          ethBalance: balance,
        },
      })
    })
  }

  networkChangedCallBack = (web3Instance, network, balance) => {
    logger.log('Network changed ', network)
    balance = web3Instance.utils.fromWei(balance, 'ether')
    this.setState({
      userData: {
        ...this.state.userData,
        currentNetwork: network,
        ethBalance: balance,
      },
    })
  }

  getUserBalance = async (web3Instance, address) => {
    let balance = 0
    if (web3Instance && address) {
      balance = await web3Instance.eth.getBalance(address)
    }
    return balance
  }

  async componentDidMount() {
    logger.log('Fire event componentDidMount')
  }

  // Converts the address from uppercase to lowercase (checksum format) in order to avoid metamask bug of using both address
  toChecksumAddress = (web3, address) => {
    if (!web3 || !address || !web3.utils) {
      return
    }
    return web3.utils.toChecksumAddress(address)
  }

  render() {
    let content = <FullLoading show={true} dismissLink={{ to: '/', text: 'Go back' }} message={this.state.displayMsg} />
    if (!this.state.requestingAuth) {
      content = (
        <Web3Context.Provider
          value={{
            web3: this.state.web3,
            connectWeb3: this.loadWeb3,
            authenticated: this.state.userData.authenticated,
            userData: this.state.userData,
            displayMsg: this.state.displayMsg,
            error: this.state.error,
          }}
        >
          {this.props.children}
        </Web3Context.Provider>
      )
    }
    return content
  }
}

export default Web3Provider
