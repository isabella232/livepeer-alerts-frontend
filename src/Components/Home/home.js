import React, { Component } from 'react'
import { ToastContainer, toast } from 'react-toastify'
import 'react-toastify/dist/ReactToastify.css'
import './home.css'

export class HomeComponent extends Component {
  state = {
    toastId: 1
  }

  onGetStartedBtnHandler = () => {
    const { web3 } = this.props
    console.log('[Home.js] props: ', this.props)
    console.log('Web3: ', web3)
    if (typeof web3 === 'undefined') {
      this.sendToastError()
    } else {
      console.log('Web3 connected ', web3.currentProvider.isConnected())
      this.props.history.push('/account')
    }
  }

  sendToastError = toastTime => {
    if (!toast.isActive(this.state.toastId)) {
      toast.error('Please install metamask and come back!', {
        position: toast.POSITION.TOP_RIGHT,
        progressClassName: 'Toast-progress-bar',
        autoClose: toastTime,
        toastId: this.state.toastId
      })
    }
  }

  componentDidMount() {
    if (!this.props.web3) {
      this.sendToastError(6000)
    }
  }

  render() {
    let content

    content = (
      <>
        <h3>Hello LivePeer Alerts!</h3>
        <p>Descriptions and details</p>
        <button onClick={this.onGetStartedBtnHandler}>Get Started</button>
        {/*      <Link to="/account">

        </Link>*/}
        <ToastContainer autoClose={3000} />
      </>
    )

    return <div>{content}</div>
  }
}
