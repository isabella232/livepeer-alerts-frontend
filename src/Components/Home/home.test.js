import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { HomeComponent } from './home'
import * as failReasons from '../Common/Hoc/Web3Provider/Web3FailReasons'
import * as texts from '../Common/UI/Texts/Texts'

configure({ adapter: new Adapter() })

describe('Renders home message', () => {
  it('Renders home message if user auth', () => {
    const message = 'Hello LivePeer Alerts!'
    let props = {
      web3: [],
      userData: {
        authenticated: true,
        reason: '',
        address: 123,
        currentNetwork: 123
      },
      render: false
    }
    let wrapper = shallow(<HomeComponent {...props} />)
    expect(wrapper.contains(message)).toEqual(true)
  })
  it('Renders error message if user not auth', () => {
    const toastFn = jest.fn()
    let props = {
      web3: [],
      userData: {
        authenticated: false,
        reason: failReasons.NO_PERMISSIONS,
      },
      render: true,
      toastOpenedHandlerTest: toastFn
    }
    let wrapper = shallow(<HomeComponent {...props} />)
    expect(toastFn).toBeCalledWith(texts.NO_PERMISSIONS)
  })
  it('Renders error message if user does not have Metamask installed', () => {
    const toastFn = jest.fn()
    let props = {
      web3: null,
      render: true,
      toastOpenedHandlerTest: toastFn,
      userData: {
        authenticated: false,
        reason: failReasons.NO_WEB3
      }
    }
    let wrapper = shallow(<HomeComponent {...props} />)
    expect(toastFn).toBeCalledWith(texts.NO_WEB3)
  })
})
