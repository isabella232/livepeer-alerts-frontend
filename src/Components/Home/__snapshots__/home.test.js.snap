// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders home message Renders home message if user auth 1`] = `
Array [
  <div
    className="HomeCard-container-27"
  >
    <div
      className="MuiGrid-container-40 MuiGrid-justify-xs-center-57 GridContainer-grid-39 undefined"
    >
      <div
        className="MuiGrid-item-41 MuiGrid-grid-xs-12-80 MuiGrid-grid-sm-12-94 MuiGrid-grid-md-10-106 GridItem-grid-137 undefined"
      >
        <div
          className="Card-card-138 HomeCard-cardSignup-28"
        >
          <h2
            className="HomeCard-cardTitle-26"
          >
            Register
          </h2>
          <div
            className="CardBody-cardBody-155"
          >
            <div
              className="MuiGrid-container-40 MuiGrid-justify-xs-center-57 GridContainer-grid-39 undefined"
            >
              <div
                className="MuiGrid-item-41 MuiGrid-grid-xs-12-80 MuiGrid-grid-sm-12-94 MuiGrid-grid-md-5-101 GridItem-grid-137 undefined"
              >
                <div
                  className="InfoArea-infoArea-164"
                >
                  <div
                    className="InfoArea-iconWrapper-165 InfoArea-rose-171"
                  >
                    <svg
                      aria-hidden="true"
                      className="MuiSvgIcon-root-177 InfoArea-icon-173"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <defs>
                        <path
                          d="M0 0h24v24H0V0z"
                          id="a"
                        />
                      </defs>
                      <path
                        d="M23 8c0 1.1-.9 2-2 2-.18 0-.35-.02-.51-.07l-3.56 3.55c.05.16.07.34.07.52 0 1.1-.9 2-2 2s-2-.9-2-2c0-.18.02-.36.07-.52l-2.55-2.55c-.16.05-.34.07-.52.07s-.36-.02-.52-.07l-4.55 4.56c.05.16.07.33.07.51 0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2c.18 0 .35.02.51.07l4.56-4.55C8.02 9.36 8 9.18 8 9c0-1.1.9-2 2-2s2 .9 2 2c0 .18-.02.36-.07.52l2.55 2.55c.16-.05.34-.07.52-.07s.36.02.52.07l3.55-3.56C19.02 8.35 19 8.18 19 8c0-1.1.9-2 2-2s2 .9 2 2z"
                      />
                    </svg>
                  </div>
                  <div
                    className="InfoArea-descriptionWrapper-174"
                  >
                    <h4
                      className="InfoArea-title-175"
                    >
                      Marketing
                    </h4>
                    <p
                      className="InfoArea-description-176"
                    >
                      We've created the marketing campaign of the website. It was a very interesting collaboration.
                    </p>
                  </div>
                </div>
                <div
                  className="InfoArea-infoArea-164"
                >
                  <div
                    className="InfoArea-iconWrapper-165 InfoArea-primary-166"
                  >
                    <svg
                      aria-hidden="true"
                      className="MuiSvgIcon-root-177 InfoArea-icon-173"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M0 0h24v24H0V0z"
                        fill="none"
                      />
                      <path
                        d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
                      />
                    </svg>
                  </div>
                  <div
                    className="InfoArea-descriptionWrapper-174"
                  >
                    <h4
                      className="InfoArea-title-175"
                    >
                      Fully Coded in HTML5
                    </h4>
                    <p
                      className="InfoArea-description-176"
                    >
                      We've developed the website with HTML5 and CSS3. The client has access to the code using GitHub.
                    </p>
                  </div>
                </div>
                <div
                  className="InfoArea-infoArea-164"
                >
                  <div
                    className="InfoArea-iconWrapper-165 InfoArea-info-170"
                  >
                    <svg
                      aria-hidden="true"
                      className="MuiSvgIcon-root-177 InfoArea-icon-173"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5c-1.66 0-3 1.34-3 3s1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5C6.34 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5z"
                      />
                    </svg>
                  </div>
                  <div
                    className="InfoArea-descriptionWrapper-174"
                  >
                    <h4
                      className="InfoArea-title-175"
                    >
                      Built Audience
                    </h4>
                    <p
                      className="InfoArea-description-176"
                    >
                      There is also a Fully Customizable CMS Admin Dashboard for this product.
                    </p>
                  </div>
                </div>
              </div>
              <div
                className="MuiGrid-item-41 MuiGrid-grid-xs-12-80 MuiGrid-grid-sm-8-90 MuiGrid-grid-md-5-101 GridItem-grid-137 undefined"
              >
                <div
                  className="HomeCard-center-29"
                  style={
                    Object {
                      "paddingBottom": "40%",
                      "paddingTop": "40%",
                    }
                  }
                >
                  <button
                    className="MuiButtonBase-root-241 MuiButton-root-215 MuiButton-text-217 MuiButton-flat-220 RegularButton-button-186 RegularButton-lg-209 RegularButton-primary-188 RegularButton-round-211"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="button"
                  >
                    <span
                      className="MuiButton-label-216"
                    >
                      Get started
                    </span>
                    <span
                      className="MuiTouchRipple-root-244"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="Toastify"
  />,
]
`;
